test/e2e/storage/testsuites/volumelimits.go

```diff
@@ -0,0 +1,335 @@
+func InitVolumeLimitsTestSuite() TestSuite {
+	return &volumeLimitsTestSuite{
+		tsInfo: TestSuiteInfo{
+			name: "volumeLimits",
+			testPatterns: []testpatterns.TestPattern{
+				testpatterns.FsVolModeDynamicPV,
+			},
+		},
+	}
+}
+
+func (t *volumeLimitsTestSuite) getTestSuiteInfo() TestSuiteInfo {
+	return t.tsInfo
+}
+
+func (t *volumeLimitsTestSuite) defineTests(driver TestDriver, pattern testpatterns.TestPattern) {
+	type local struct {
+		config      *PerTestConfig
+		testCleanup func()
+
+		cs clientset.Interface
+		ns *v1.Namespace
+		// genericVolumeTestResource contains pv, pvc, sc, etc. of the first pod created
+		resource *genericVolumeTestResource
+
+		// All created PVCs, incl. the one in resource
+		pvcs []*v1.PersistentVolumeClaim
+
+		// All created PVs, incl. the one in resource
+		pvNames sets.String
+
+		runningPod       *v1.Pod
+		unschedulablePod *v1.Pod
+	}
+	var (
+		l local
+	)
+
+	// No preconditions to test. Normally they would be in a BeforeEach here.
+	f := framework.NewDefaultFramework("volumelimits")
+
+	// This checks that CSIMaxVolumeLimitChecker works as expected.
+	// A randomly chosen node should be able to handle as many CSI volumes as
+	// it claims to handle in CSINode.Spec.Drivers[x].Allocatable.
+	// The test uses one single pod with a lot of volumes to work around any
+	// max pod limit on a node.
+	// And one extra pod with a CSI volume should get Pending with a condition
+	// that says it's unschedulable because of volume limit.
+	// BEWARE: the test may create lot of volumes and it's really slow.
+	ginkgo.It("should support volume limits [Slow][Serial]", func() {
+		driverInfo := driver.GetDriverInfo()
+		if !driverInfo.Capabilities[CapVolumeLimits] {
+			ginkgo.Skip(fmt.Sprintf("driver %s does not support volume limits", driverInfo.Name))
+		}
+		var dDriver DynamicPVTestDriver
+		if dDriver = driver.(DynamicPVTestDriver); dDriver == nil {
+			ginkgo.Fail("Test driver does not provide dynamically created volumes")
+		}
+
+		l.ns = f.Namespace
+		l.cs = f.ClientSet
+		l.config, l.testCleanup = driver.PrepareTest(f)
+		defer l.testCleanup()
+
+		ginkgo.By("Picking a random node")
+		var nodeName string
+		nodeList := framework.GetReadySchedulableNodesOrDie(f.ClientSet)
+		if len(nodeList.Items) != 0 {
+			nodeName = nodeList.Items[0].Name
+		} else {
+			e2elog.Failf("Unable to find ready and schedulable Node")
+		}
+		e2elog.Logf("Selected node %s", nodeName)
+
+		ginkgo.By("Checking node limits")
+		limit, err := getNodeLimits(l.cs, nodeName, driverInfo)
+		framework.ExpectNoError(err)
+
+		e2elog.Logf("Node %s can handle %d volumes of driver %s", nodeName, limit, driverInfo.Name)
+		// Create a storage class and generate a PVC. Do not instantiate the PVC yet, keep it for the last pod.
+		l.resource = createGenericVolumeTestResource(driver, l.config, pattern)
+		defer l.resource.cleanupResource()
+
+		// Prepare cleanup
+		defer func() {
```

Comment:
maybe split some of this cleanup function into helper functions, it's very long and complicated right now

Reply:
Moved to separate func.